File: Contents/27/27_winio_server.cs
using System;
using System.IO.Pipes;
using System.IO;

public class NamedPipe
{
	[STAThread]
	public static void Main(string[] args)
	{
		// 名前付きパイプを作成
		using (NamedPipeServerStream pipeServer = new NamedPipeServerStream("mypipe", PipeDirection.InOut))
		{
			Console.WriteLine("C# server is listening...");

			// クライアントの接続を待機
			pipeServer.WaitForConnection();

			Console.WriteLine("Client connected.");

			using (StreamReader reader = new StreamReader(pipeServer))
			using (StreamWriter writer = new StreamWriter(pipeServer) { AutoFlush = true })
			{
				// 受信
				string line = reader.ReadLine();
				Console.WriteLine("Received from Go: " + line);
				
				// 送信
				writer.WriteLine("Hello from C#");

				// 受信
				line = reader.ReadLine();
				Console.WriteLine("Received from Go: " + line);
			}
		}
	}
}

File: Contents/27/27_winio_server.go
package main

import (
	"bufio"
	"fmt"
	"log"

	"github.com/Microsoft/go-winio"
)

const pipeName = `\\.\pipe\mypipe`

func main() {
	// 名前付きパイプのリスナーを作成
	listener, err := winio.ListenPipe(pipeName, nil)
	if err != nil {
		log.Fatal(err)
	}
	defer listener.Close()

	fmt.Println("Go server is listening...")

	// クライアントからの接続を待機
	conn, err := listener.Accept()
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	fmt.Println("Client connected.")

	// 受信
	scanner := bufio.NewScanner(conn)
	for scanner.Scan() {
		fmt.Println("Received from C#: " + scanner.Text())
		break
	}

	// 送信
	fmt.Fprintln(conn, "Hello from Go")

	// 受信
	scanner = bufio.NewScanner(conn)
	for scanner.Scan() {
		fmt.Println("Received from C#: " + scanner.Text())
		break
	}
}
